src/toxcore/dht/dht_friend.rs

```diff
@@ -57,54 +57,43 @@ impl DhtFriend {
     }
 
     /// get Socket Address list of a friend, a friend can have multi IP address bacause of NAT
-    pub fn get_addrs_of_clients(&self, is_ipv6_mode: bool) -> Vec<SocketAddr> {
+    pub fn get_addrs_of_clients(&self) -> Vec<SocketAddr> {
+        let mut socks = Vec::new();
+        let mut direct_connected = false;
+
         self.close_nodes.nodes.iter()
-            .map(|node| node.get_socket_addr(is_ipv6_mode))
-            .filter_map(|addr| addr)
-            .collect::<Vec<SocketAddr>>()
+            .for_each(|node| {
+                if let Some(v6) = node.assoc6.ret_saddr {
+                    if !node.assoc6.is_bad() {
+                        socks.push(SocketAddr::V6(v6));
+                    }
+                }
+
+                if let Some(v4) = node.assoc4.ret_saddr {
+                    if !node.assoc4.is_bad() {
+                        socks.push(SocketAddr::V4(v4));
+                    }
+                }
+
+                if self.pk == node.pk && (!node.assoc4.is_bad() || !node.assoc6.is_bad()) {
+                    direct_connected = true;
```

Comment:
No need in this variable. You can return empty vector right here.

Reply:
Is it possible to break for_each()?

Reply:
Sure.